Unary Stencil

dpvanbalen opened the pull request:
**Description**
Adds the option to have a unary dimension in a stencil.

**Motivation and context**
Cleans up the interface a bit, and using such a unary dimension should also make the generated code faster.
There is one open question: Currently, I just use `e` as the unary stencil of 1 `e`s. Alternatively, we could use something like https://hackage.haskell.org/package/base-4.16.0.0/docs/Data-Tuple.html#t:Solo.

**How has this been tested?**
Tests pass

**Types of changes**
What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Checklist**
Go over all the following points, and put an `x` in all the boxes that apply. If you're unsure about any of these, don't hesitate to ask. We're here to help!

- [ ] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

dpvanbalen commented:
Looking again, perhaps it is better to go with the `Solo` route: The example in the haddoc is not run by the doctest, and trying myself I get a different answer. I assume that it interprets both stencils to be horizontal.
Also, I just noticed that Data.Array.Accelerate.Sugar.Stencil _also_ exists. I didn't add my instances there, and nothing complains.. is it relevant?

dpvanbalen commented:
This version now uses `Solo`, and now this example works as expected:
```
  let
    convolve5x1 :: [A.Exp Double] -> A.Stencil5x1 Double -> A.Exp Double
    convolve5x1 kernel (A.Solo (a,b,c,d,e))
      = Prelude.sum $ Prelude.zipWith (*) kernel [a,b,c,d,e]
   
    convolve1x5 :: [A.Exp Double] -> A.Stencil1x5 Double -> A.Exp Double
    convolve1x5 kernel (A.Solo a, A.Solo b, A.Solo c, A.Solo d, A.Solo e)
      = Prelude.sum $ Prelude.zipWith (*) kernel [a,b,c,d,e]
    gaussian :: [A.Exp Double]
    gaussian = [0.06136,0.24477,0.38774,0.24477,0.06136]

  in   A.stencil (convolve5x1 gaussian) A.clamp
     . A.stencil (convolve1x5 gaussian) A.clamp
```
We could also use anything other than Solo, or e.g. also export `pattern T0 :: Exp ()` `pattern T1 :: Exp a -> Exp (Solo a)`

tmcdonell commented:
Good question r.e. Sugar.Stencil... I'm guessing that that code used to live in Smart and... *checks* wait, that code still exists in Smart. Uh, I think that was not properly cleaned up in the conversion to representation types?

Needing to use `Solo` is unfortunate, is it better than what we have now? I not sure but I guess so?

tmcdonell commented:
Also there are currently no tests for `Nx1` or `1xN` stencils, so we should add those (:

dpvanbalen commented:
I changed `Solo` to `newtype Unary a = a`, because GHC.Tuple doesn't export Solo for GHC<9.
Also added some tests, and moved the stencil logic from Smart to Sugar.Stencil.

I agree that it would be optimal if we didn't need Unary or Solo, but it seems like the only way to make the types work out. It is probably still a more sensible idiom than (_, a, _), especially because this also has performance benefits!

tomsmeding commented:
Regarding `Solo` not being available on older GHCs: there is a compatibility package [OneTuple](https://hackage.haskell.org/package/OneTuple-0.4.1.1/docs/Data-Tuple-Solo.html) on Hackage that re-exports GHC's own Solo for new enough GHCs. I think that would give the best compatibility story.

tomsmeding commented on src/Data/Array/Accelerate/Language.hs:
```diff
@@ -97,7 +97,7 @@ module Data.Array.Accelerate.Language (
   ord, chr, boolToInt, bitcast,
 
-) where
+Stencil1,Stencil1x3,Stencil1x5,Stencil1x1,Stencil3x1,Stencil5x1) where
```
A newline before the `)` and an indent before the names would be nice here

tomsmeding commented on src/Data/Array/Accelerate/Language.hs:
```diff
@@ -117,8 +117,8 @@ import Data.Array.Accelerate.Classes.Fractional
 import Data.Array.Accelerate.Classes.Ord
-
 import Prelude                                                      ( ($), (.), Maybe(..), Char )
+import Data.Array.Accelerate.Sugar.Stencil
```
The imports were structured before -- whether that's useful is debatable. But I think `import Prelude` should be separate because it should catch the eye.

tomsmeding commented on src/Data/Array/Accelerate/Representation/Type.hs:
```diff
@@ -123,5 +123,5 @@ runQ $
          tySynD (mkName ("Tup" ++ show n)) (map plainTV xs) rhs
   in
-  mapM mkT [2..16]
+  mapM mkT [0..16]
```
Zero as well?

tomsmeding reviewed (commented):
Some minor things.
Does this need a PR on `accelerate-llvm` too?
